enterprise/backend/src/metabase_enterprise/database_replication/api.clj

```diff
@@ -31,44 +31,55 @@
-                                      (some #(re-matches % schema) include-patterns)
-                                      (not-empty exclude-patterns)
-                                      (not (some #(re-matches % schema) exclude-patterns)))))
-                          (map :estimated_row_count)
-                          ;; FIXME: `estimated_row_count` might be `nil`, in which case we should tell the user:
-                          ;;  "we don't know yet whether this will work, wait or try your luck"
-                          (reduce +))
-                         0)]
-    (log/infof "Quota left: %s. Estimate db row count: %s" free-quota total-row-count)
-    (< total-row-count free-quota)))
+  (let [free-quota                (or
+                                   (some->> (m/find-first (comp #{"clickhouse-dwh"} :hosting-feature) (quotas))
+                                            ((juxt :soft-limit :usage))
+                                            (apply -))
+                                   -1)
+        include-patterns          (->> schemas
+                                       (filter (comp #{"include"} :type))
+                                       (map (comp re-pattern :pattern)))
+        exclude-patterns          (->> schemas
+                                       (filter (comp #{"exclude"} :type))
+                                       (map (comp re-pattern :pattern)))
+        total-estimated-row-count (or
+                                   (some->>
+                                    (t2/hydrate database [:tables :fields])
+                                    :tables
+                                    (filter (comp (fn [x] (re-matches #"^[A-Za-z0-9_]+$" x)) :name))       ; sanitized name
+                                    (filter (fn [t] (some (comp #{:type/PK} :semantic_type) (:fields t)))) ; has-pkey
+                                    (filter (fn [{:keys [schema]}]
+                                              (cond
+                                                (empty? schemas)
+                                                true
+                                                (not-empty include-patterns)
+                                                (some #(re-matches % schema) include-patterns)
+                                                (not-empty exclude-patterns)
+                                                (not (some #(re-matches % schema) exclude-patterns)))))
+                                    (map :estimated_row_count)
+                                    ;; FIXME: `estimated_row_count` might be `nil`, in which case we should tell the user:
+                                    ;;  "we don't know yet whether this will work, wait or try your luck"
+                                    (reduce +))
+                                   0)]
+    (log/infof "Quota left: %s. Estimate db row count: %s" free-quota total-estimated-row-count)
+    {:free-quota                free-quota
```

Comment:
Should be include `quotas` here?  The limits and usage might be helpful in the modal.

Reply:
yup! I think we can do the next part on top of this commit (no need to backnforth)